beacon-chain/blockchain/core.go

```diff
@@ -480,7 +492,68 @@ func (b *BeaconChain) initCycle(cState *types.CrystallizedState, aState *types.A
 		RecentBlockHashes:   recentBlockHashes,
 	}, aState.GetBlockVoteCache())
 
-	return newCrystallizedState, newActiveState
+	return newCrystallizedState, newActiveState, nil
+}
+
+// processCrosslinks checks if the proposed shard block has recevied
+// 2/3 of the votes. If yes, we update crosslink record to point to
+// the proposed shard block with latest dynasty and slot numbers.
+func (b *BeaconChain) processCrosslinks(
+	crosslinkRecords []*pb.CrosslinkRecord,
+	validators []*pb.ValidatorRecord,
+	pendingAttestations []*pb.AttestationRecord,
+	dynasty uint64,
+	slot uint64) ([]*pb.CrosslinkRecord, error) {
+
+	type shardBlockHash struct {
+		id        uint64
+		blockHash []byte
+	}
+	shardBlockHashBalance := make(map[*shardBlockHash]uint64)
+
+	for _, attestation := range pendingAttestations {
+		sbh := shardBlockHash{
+			id:        attestation.ShardId,
+			blockHash: attestation.ShardBlockHash,
+		}
+		var keyExists bool
+		if _, ok := shardBlockHashBalance[&sbh]; ok {
+			keyExists = true
+		}
+		if !keyExists {
+			shardBlockHashBalance[&sbh] = 0
+		}
+
+		indices, err := b.getAttesterIndices(attestation)
+		if err != nil {
+			return nil, err
```

Comment:
We should also remember to not modify the original state. Notice how if we error out halfway through, we exit out of `stateRecalc` but our crystallized state will be polluted with unrelated crosslinks.